### PR TITLE
Encryption: seprate the cipher method from AESCTRCipherStream to support encrypt/decrypt simple data

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #8351 

Problem Summary:

### What is changed and how it works?

To support this:

```cpp
String key = "random-key";
String data = "random-string";
Cipher(0, data.data(), data.size(), key, EncryptionMethod::Aes256Ctr, nullptr, true);
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Encryption/AESCTRCipher.cpp [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        // Use sm4 in GmSSL, return nullptr
+        return nullptr;
+#elif OPENSSL_VERSION_NUMBER < 0x1010100fL || defined(OPENSSL_NO_SM4)
+        throw DB::TiFlashException(
+            Errors::Encryption::Internal,
+            "Unsupported encryption method: {}",
+            static_cast<int>(method));
+#else
+        // Openssl support SM4 after 1.1.1 release version.
+        cipher = EVP_sm4_ctr();
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Encryption/AESCTRCipher.cpp]: fixed

File: dbms/src/Encryption/AESCTRCipher.cpp [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    case EncryptionMethod::SM4Ctr:
+#if USE_GM_SSL
+        // SM4 uses 128-bit block size as AES.
+        // Ref: https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/include/crypto/sm4.h#L24
+        return 16;
+#elif OPENSSL_VERSION_NUMBER < 0x1010100fL || defined(OPENSSL_NO_SM4)
+        throw DB::TiFlashException(
+            Errors::Encryption::Internal,
+            "Unsupported encryption method: {}",
+            static_cast<int>(method));
```
Comment (CalvinNeo) [inline dbms/src/Encryption/AESCTRCipher.cpp:61]: Can we use magic enum here?

File: dbms/src/Encryption/AESCTRCipher.cpp [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        return EVP_aes_256_ctr();
+    case EncryptionMethod::SM4Ctr:
+#if USE_GM_SSL
+        // Use sm4 in GmSSL, return nullptr
+        return nullptr;
+#elif OPENSSL_VERSION_NUMBER < 0x1010100fL || defined(OPENSSL_NO_SM4)
+        throw DB::TiFlashException(
+            Errors::Encryption::Internal,
+            "Unsupported encryption method: {}",
+            static_cast<int>(method));
```
Comment (CalvinNeo) [inline dbms/src/Encryption/AESCTRCipher.cpp:89]: ditto

File: dbms/src/Encryption/AESCTRCipher.cpp (hunk 2 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Encryption/AESCTRCipher.cpp]: fixed

File: dbms/src/Encryption/AESCTRCipher.cpp [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        // Ref: https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/include/crypto/sm4.h#L24
+        return 16;
+#elif OPENSSL_VERSION_NUMBER < 0x1010100fL || defined(OPENSSL_NO_SM4)
+        throw DB::TiFlashException(
+            Errors::Encryption::Internal,
+            "Do not support encryption method SM4Ctr when using OpenSSL and the version is less than 1.1.1");
+#else
+        // Openssl support SM4 after 1.1.1 release version.
+        return SM4_BLOCK_SIZE;
+#endif
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/AESCTRCipher.cpp:64]: why not keep the previous definition?
```
#if defined(SM4_BLOCK_SIZE)
        return SM4_BLOCK_SIZE;
#else
        throw DB::TiFlashException(...
#endif
```

Comment (Lloyd-Pottiger) [reply dbms/src/Encryption/AESCTRCipher.cpp:64]:
```
#if (USE_GM_SSL == 0) && !defined(OPENSSL_NO_SM4)
// TODO: OpenSSL Lib does not export SM4_BLOCK_SIZE by now.
// Need to remove SM4_BLOCK_SIZE once Openssl lib support the definition.
// SM4 uses 128-bit block size as AES.
// Ref:
// https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/include/crypto/sm4.h#L24
#define SM4_BLOCK_SIZE 16
#endif
```

combine with this, make it easy to understand

File: dbms/src/Encryption/AESCTRCipher.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+#endif
+
+#if (USE_GM_SSL == 0) && !defined(OPENSSL_NO_SM4)
+// TODO: OpenSSL Lib does not export SM4_BLOCK_SIZE by now.
+// Need to remove SM4_BLOCK_SIZE once Openssl lib support the definition.
+// SM4 uses 128-bit block size as AES.
+// Ref:
+// https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/include/crypto/sm4.h#L24
+#define SM4_BLOCK_SIZE 16
+#endif
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/AESCTRCipher.h:59]: I think we can move these into the `.cpp` file

File: dbms/src/Encryption/AESCTRCipherStream.cpp [hunk 6]
```diff
@@ -29,253 +30,50 @@ namespace ErrorCodes
...
-                "Unexpected cipher output size for last block, expected {} actual {}",
-                block_size,
-                output_size);
-#if USE_GM_SSL
-        }
-#endif
-        memcpy(data + data_offset, partial_block, remaining_data_size);
-    }
-#endif
+    DB::Encryption::Cipher(file_offset, data, data_size, key, method, iv, /*is_encrypt=*/true);
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/AESCTRCipherStream.cpp:55]: In the master implementation, `AESCTRCipherStream` use the same `cipher` instance for all `encryption` and `decryption` calls. But now it create a new `cipher` instance for each `encryption` and `decryption` calls.
Will it bring different encryption result? Will it bring performance issue?

Comment (Lloyd-Pottiger) [reply dbms/src/Encryption/AESCTRCipherStream.cpp:55]: It just return a pointer which point to a static instance, no new instance is created.

File: dbms/src/Encryption/AESCTRCipher.h (hunk 5 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Encryption/AESCTRCipher.h]: done

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Review by CalvinNeo — APPROVED** [review]
lgtm

**Comment by JaySon-Huang** [comment]
```
[2023-12-19T02:11:23.943Z] tidb-ci/fullstack-test-dt/union_push_down.test: Running
[2023-12-19T02:11:29.477Z]   File: tidb-ci/fullstack-test-dt/union_push_down.test
[2023-12-19T02:11:29.477Z]   Error line: 29
[2023-12-19T02:11:29.477Z]   Error: analyze table test.tt;
[2023-12-19T02:11:29.477Z]   Result:
[2023-12-19T02:11:29.477Z]     ERROR 1105 (HY000) at line 1: rpc error: code = Unknown desc = [PD:tso:ErrGenerateTimestamp]generate timestamp failed, requested pd is not leader of cluster
[2023-12-19T02:11:29.477Z]   Expected:
[2023-12-19T02:11:29.477Z]     <nothing>
```

**Comment by Lloyd-Pottiger** [comment]
```
[2023-12-19T02:11:23.943Z] tidb-ci/fullstack-test-dt/union_push_down.test: Running
[2023-12-19T02:11:29.477Z]   File: tidb-ci/fullstack-test-dt/union_push_down.test
[2023-12-19T02:11:29.477Z]   Error line: 29
[2023-12-19T02:11:29.477Z]   Error: analyze table test.tt;
[2023-12-19T02:11:29.477Z]   Result:
[2023-12-19T02:11:29.477Z]     ERROR 1105 (HY000) at line 1: rpc error: code = Unknown desc = [PD:tso:ErrGenerateTimestamp]generate timestamp failed, requested pd is not leader of cluster
[2023-12-19T02:11:29.477Z]   Expected:
[2023-12-19T02:11:29.477Z]     <nothing>
```

**Comment by JaySon-Huang** [comment]
https://ci.pingcap.net/blue/organizations/jenkins/tiflash-ghpr-integration-tests/detail/tiflash-ghpr-integration-tests/15506/pipeline/128
```
[2023-12-19T02:57:22.328Z] [2023/12/19 10:56:58.600 +08:00] [ERROR] [<unknown>] ["Receive TsoResponse failed"] [source=pingcap.pd] [thread_id=1191]
[2023-12-19T02:57:22.328Z] [2023/12/19 10:56:58.602 +08:00] [WARN] [<unknown>] ["update ts error: Exception: Receive TsoResponse failed"] [source=pd/oracle] [thread_id=1191]
[2023-12-19T02:57:22.328Z] [2023/12/19 10:56:58.719 +08:00] [ERROR] [<unknown>] ["Receive TsoResponse failed"] [source=pingcap.pd] [thread_id=302]
[2023-12-19T02:57:22.328Z] [2023/12/19 10:56:58.720 +08:00] [FATAL] [Exception.cpp:106] ["Poco::Exception. Code: 1000, e.code() = 2, e.displayText() = Exception: Receive TsoResponse failed, e.what() = Exception"] [source="DB::RawCppPtr DB::PreHandleSnapshot(DB::EngineStoreServerWrap *, DB::BaseBuffView, uint64_t, DB::SSTViewVec, uint64_t, uint64_t)"] [thread_id=302]
```

**Comment by Lloyd-Pottiger** [comment]
after

```
[Summary] Q1: 1.78s
[Summary] Q10: 1.95s
[Summary] Q11: 0.58s
[Summary] Q12: 1.06s
[Summary] Q13: 2.04s
[Summary] Q14: 0.58s
[Summary] Q15: 1.28s
[Summary] Q16: 0.51s
[Summary] Q17: 4.19s
[Summary] Q18: 3.84s
[Summary] Q19: 0.94s
[Summary] Q2: 0.86s
[Summary] Q20: 0.90s
[Summary] Q21: 3.59s
[Summary] Q22: 0.37s
[Summary] Q3: 1.55s
[Summary] Q4: 4.63s
[Summary] Q5: 2.83s
[Summary] Q6: 0.45s
[Summary] Q7: 1.64s
[Summary] Q8: 2.32s
[Summary] Q9: 8.12s
```

**Comment by Lloyd-Pottiger** [comment]
/unhold

No performance regression found now, maybe due to the CI jitter, so merge this pr.

**Pull request merged**